src/celeritas/optical/surface/SurfaceTraversalView.hh

```diff
@@ -0,0 +1,235 @@
+ */
+CELER_FUNCTION
+SurfaceTraversalView::SurfaceTraversalView(SurfaceParamsRef const& params,
+                                           SurfaceStateRef const& states,
+                                           TrackSlotId track)
+    : params_(params), states_(states), track_id_(track)
+{
+    CELER_EXPECT(track_id_ < states_.size());
+    CELER_EXPECT(states_.surface[track_id_] < params_.surfaces.size());
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Initialize track state with given initializer data.
+ */
+CELER_FUNCTION SurfaceTraversalView&
+SurfaceTraversalView::operator=(Initializer const&)
+{
+    states_.surface_position[track_id_] = SurfaceTrackPosition{0};
+    states_.track_direction[track_id_] = SubsurfaceDirection::forward;
+    return *this;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Whether the track is in the pre-volume.
+ */
+CELER_FUNCTION bool SurfaceTraversalView::in_pre_volume() const
+{
+    return this->position().get() == 0;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Whether the track is in the post-volume.
+ */
+CELER_FUNCTION bool SurfaceTraversalView::in_post_volume() const
+{
+    return this->position().get() + 1 == this->num_positions();
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Whether the current track direction is exiting the surface.
+ */
+CELER_FUNCTION bool SurfaceTraversalView::is_exiting() const
+{
+    return this->is_exiting(this->direction());
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Whether the direction is exiting the surface.
+ */
+CELER_FUNCTION bool
+SurfaceTraversalView::is_exiting(SubsurfaceDirection d) const
+{
+    // Use unsigned underflow when moving reverse (-1) on the pre-surface
+    // (position 0) to wrap to an invalid position value
+    return advance_subsurface_position_along(this->position(), d).unchecked_get()
+           >= this->num_positions();
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Current position of the track in the sub-surfaces, in track-local
+ * coordinates.
+ *
+ * Tracks traverse a surface in track-local coordinates where 0 is the
+ * pre-volume and N is the post-volume. Depending on the surface orientation,
+ * this will be mapped to the appropriate sub-surface material and interface.
+ */
+CELER_FUNCTION SurfaceTrackPosition SurfaceTraversalView::position() const
+{
+    return states_.surface_position[track_id_];
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ */
+CELER_FUNCTION SurfaceTrackPosition SurfaceTraversalView::next_position() const
+{
+    return advance_subsurface_position_along(this->position(),
+                                             this->direction());
+}
```

Comment:
Another good candidate for a free function: `SurfaceTrackPosition  next_position(SurfaceTraversalView const&)` (and in fact rename `advance_subsurface_position_along`)